cdap-tms/src/main/java/co/cask/cdap/messaging/store/hbase/HBaseMessageTable.java

```diff
@@ -0,0 +1,234 @@
+import co.cask.cdap.messaging.store.leveldb.MessageEntry;
+import co.cask.cdap.proto.id.TopicId;
+import com.google.common.base.Throwables;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.client.Delete;
+import org.apache.hadoop.hbase.client.HTable;
+import org.apache.hadoop.hbase.client.Put;
+import org.apache.hadoop.hbase.client.Result;
+import org.apache.hadoop.hbase.client.ResultScanner;
+import org.apache.hadoop.hbase.client.Scan;
+import org.apache.tephra.Transaction;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * HBase implementation of {@link MessageTable}.
+ */
+public class HBaseMessageTable implements MessageTable {
+  private static final byte[] PAYLOAD_COL = Bytes.toBytes('p');
+  private static final byte[] TX_COL = Bytes.toBytes('t');
+  private final Configuration hConf;
+  private final HBaseTableUtil tableUtil;
+  private final TableId tableId;
+  private final byte[] columnFamily;
+  private volatile long writeTimestamp;
+  private volatile short seqId;
+  private volatile HTable hTable;
+
+  public HBaseMessageTable(Configuration hConf, HBaseTableUtil tableUtil, TableId tableId, byte[] columnFamily) {
+    this.hConf = hConf;
+    this.tableUtil = tableUtil;
+    this.tableId = tableId;
+    this.columnFamily = Arrays.copyOf(columnFamily, columnFamily.length);
+  }
+
+  @Override
+  public CloseableIterator<Entry> fetch(TopicId topicId, long startTime, int limit, @Nullable Transaction transaction)
+    throws IOException {
+    byte[] topic = MessagingUtils.toBytes(topicId);
+    byte[] startRow = Bytes.add(topic, Bytes.toBytes(startTime));
+    Scan scan = tableUtil.buildScan()
+      .setStartRow(startRow)
+      .setStopRow(Bytes.stopKeyForPrefix(topic))
+      .build();
+    ResultScanner scanner = getHTable().getScanner(scan);
+    return new HBaseCloseableIterator(scanner, limit, columnFamily, transaction);
+  }
+
+  @Override
+  public CloseableIterator<Entry> fetch(TopicId topicId, MessageId messageId, boolean inclusive,
+                                        int limit, @Nullable Transaction transaction) throws IOException {
+    byte[] topic = MessagingUtils.toBytes(topicId);
+    byte[] startRow = Bytes.add(topic, Bytes.toBytes(messageId.getPublishTimestamp()),
+                                Bytes.toBytes(messageId.getSequenceId()));
+    if (!inclusive) {
+      startRow = Bytes.incrementBytes(startRow, 1);
+    }
+    Scan scan = tableUtil.buildScan()
+      .setStartRow(startRow)
+      .setStopRow(Bytes.stopKeyForPrefix(topic))
+      .build();
+    ResultScanner scanner = getHTable().getScanner(scan);
+    return new HBaseCloseableIterator(scanner, limit, columnFamily, transaction);
+  }
+
+  @Override
+  public void store(Iterator<Entry> entries) throws IOException {
+    long writeTs = System.currentTimeMillis();
```

Comment:
Just a note, I am adding a class `TimeProvider` to cdap-common so that we can switch to use a different time generator for easy testing purpose. We can switch to it after that is merged.